### PR TITLE
[POAE7-2749] [Hotfix] Fix Wrong Function Declaration Copy Procedure in cloneFunctionRecursive()

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Fix Wrong Function Declaration Copy Procedure in cloneFunctionRecursive()

### Why are the changes needed?

BugFix

### Which label does this PR belong to?

BugFix

Pls take a review @kd97 @yma11

### CODE REVIEW

**Review by yma11 — APPROVED** [review]
I verified it in my scenario and it works. But logic is weird here and seems we actually disable the recursive clone since in most cases `fn->isDeclaration()` is true?

**Comment by bigPYJ1151** [comment]
> I verified it in my scenario and it works. But logic is weird here and seems we actually disable the recursive clone since in most cases `fn->isDeclaration()` is true?

Emm, you're right, seems it not the root cause.

**Review by kd97 — DISMISSED** [review]

**Comment by bigPYJ1151** [comment]
Pls help to check, thx. @yma11 @kd97

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp [hunk 1]
```diff
@@ -254,7 +255,11 @@ void LLVMJITFunction::cloneFunctionRecursive(llvm::Function* fn) {
   for (auto it = llvm::inst_begin(fn), e = llvm::inst_end(fn); it != e; ++it) {
     if (llvm::isa<llvm::CallInst>(*it)) {
       auto& call = llvm::cast<llvm::CallInst>(*it);
-      cloneFunctionRecursive(call.getCalledFunction());
+      llvm::Function* called_func = call.getCalledFunction();
+      if (called_func && called_func->hasFnAttribute(llvm::Attribute::AlwaysInline)) {
+        // Only symbol-called function with always-inline atrribute needs to clone.
+        cloneFunctionRecursive(called_func);
+      }
     }
```
Comment (kd97) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:263]: what determines the `AlwaysInline` attribute (eg, func definition) ? And no need to clone if not go through this branch ?

Comment (kd97) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:263]: seems AlwaysInline equals recursion termination condition？

Comment (bigPYJ1151) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:263]: Runtime function defined with ```ALWAYSINLINE``` has ```AlwaysInline``` attribute. The only reason to clone function definition is inlining the function, or function declaration is enough.

Comment (yma11) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:263]: So for the runtime function which called object's API should not use `ALWAYS_INLINE` label? like call `ordered_set::insert()`.

Comment (bigPYJ1151) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:263]: > So for the runtime function which called object's API should not use `ALWAYS_INLINE` label? like call `ordered_set::insert()`.

No such restriction.

**Review by yma11 — APPROVED** [review]

**Review by kd97 — APPROVED** [review]

**Pull request merged**